ffi/src/lib.rs

```diff
@@ -247,17 +255,737 @@ pub extern "C" fn store_open(uri: *const c_char) -> *mut Store {
+/// allow for programmatic assertion of values for that entity.
+///
+/// # Safety
+///
+/// Callers are responsible for managing the memory for the return value.
+/// A destructor `entity_builder_destroy` is provided for releasing the memory for this
+/// pointer type.
+#[no_mangle]
+pub unsafe extern "C" fn in_progress_entity_builder_from_entid<'m>(in_progress: *mut InProgress<'m, 'm>, entid: c_longlong) -> *mut EntityBuilder<InProgressBuilder> {
+    let in_progress = Box::from_raw(in_progress);
+    Box::into_raw(Box::new(in_progress.builder().describe(&KnownEntid(entid))))
+}
+
+/// Starts a new transaction and creates a builder using the that transaction
+/// to allow for programmatic assertion of values.
+///
+/// # Safety
+///
+/// Callers are responsible for managing the memory for the return value.
+/// A destructor `in_progress_builder_destroy` is provided for releasing the memory for this
+/// pointer type.
+#[no_mangle]
+pub unsafe extern "C" fn store_in_progress_builder(store: *mut Store) -> *mut ExternResult {
+    let store = &mut*store;
+    let result = store.begin_transaction().and_then(|in_progress| {
+        Ok(in_progress.builder())
+    });
+    Box::into_raw(Box::new(result.into()))
+}
+
+/// Starts a new transaction and creates a builder for an entity with `tempid`
+/// using the that transaction to allow for programmatic assertion of values for that entity.
+///
+/// # Safety
+///
+/// Callers are responsible for managing the memory for the return value.
+/// A destructor `entity_builder_destroy` is provided for releasing the memory for this
+/// pointer type.
+#[no_mangle]
+pub unsafe extern "C" fn store_entity_builder_from_temp_id(store: *mut Store, temp_id: *const c_char) -> *mut ExternResult {
+    let store = &mut*store;
+    let temp_id = c_char_to_string(temp_id);
+    let result = store.begin_transaction().and_then(|in_progress| {
+        Ok(in_progress.builder().describe_tempid(&temp_id))
+    });
+    Box::into_raw(Box::new(result.into()))
+}
+
+/// Starts a new transaction and creates a builder for an entity with `entid`
+/// using the that transaction to allow for programmatic assertion of values for that entity.
+///
+/// # Safety
+///
+/// Callers are responsible for managing the memory for the return value.
+/// A destructor `entity_builder_destroy` is provided for releasing the memory for this
 /// pointer type.
+#[no_mangle]
+pub unsafe extern "C" fn store_entity_builder_from_entid(store: *mut Store, entid: c_longlong) -> *mut ExternResult {
+    let store = &mut*store;
+    let result = store.begin_transaction().and_then(|in_progress| {
+        Ok(in_progress.builder().describe(&KnownEntid(entid)))
+    });
+    Box::into_raw(Box::new(result.into()))
+}
+
+/// Uses `builder` to assert `value` for `kw` on entity `entid`.
+///
+/// # Errors
+///
+/// If `entid` is not present in the store.
+/// If `kw` is not a valid attribute in the store.
+/// If the type of `kw` is not `:db.type/string`.
```

Comment:
This type is a schema-level consideration; it doesn't map directly to a concrete type in Rust, and it definitely doesn't in the FFI.  Do you mean it should be `TypedValue::String`?  I don't think so since you're interpreting it as a `char*`.  And even then you're turning it into a `TypedValue::Keyword`... did you mean `value` should be a string?

Reply:
Yes, I think you did mean `value` should be of a certain type, here and many other places.  But the notes on the type still apply -- you either want a real, Rust, concrete type; or you want to talk about coercion into the value space in whatever way makes sense for FFI consumers.

Reply:
Changed to ```If the `:db/type` of the attribute described by `kw` is not``` instead